File: cookbook/python/11_template_management.py
#!/usr/bin/env python3
"""
HOPX.AI Python SDK - Template Management

This example demonstrates:
1. Listing available templates
2. Getting template details and checking status
3. Creating sandboxes from templates
"""

import os
from hopx_ai import Sandbox

# Get API key from environment
API_KEY = os.environ.get('HOPX_API_KEY')
if not API_KEY:
    raise ValueError("HOPX_API_KEY environment variable not set")


def list_all_templates():
    """List all available templates"""
    print("="*60)
    print("1. LIST ALL TEMPLATES")
    print("="*60)
    
    templates = Sandbox.list_templates(api_key=API_KEY)
    
    print(f"Found {len(templates)} template(s):\n")
    
    for template in templates:
        print(f"📦 {template.name}")
        print(f"   ID: {template.id}")
        print(f"   Display Name: {template.display_name}")
        if hasattr(template, 'status') and template.status:
            print(f"   Status: {template.status}")
        print(f"   Is Active: {template.is_active}")
        if hasattr(template, 'category') and template.category:
            print(f"   Category: {template.category}")
        if hasattr(template, 'description') and template.description:
            print(f"   Description: {template.description}")
        print()
    
    return templates


def get_template_details():
    """Get details for a specific template"""
    print("="*60)
    print("2. GET TEMPLATE DETAILS")
    print("="*60)
    
    template_name = "code-interpreter"
    print(f"Getting details for '{template_name}'...\n")
    
    template = Sandbox.get_template(template_name, api_key=API_KEY)
    
    print(f"✅ Template: {template.name}")
    print(f"   Display Name: {template.display_name}")
    if hasattr(template, 'status') and template.status:
        print(f"   Status: {template.status}")
    print(f"   Is Active: {template.is_active}")
    if hasattr(template, 'category') and template.category:
        print(f"   Category: {template.category}")
    if hasattr(template, 'description') and template.description:
        print(f"   Description: {template.description}")
    print()
    
    # Important: Check status before using template
    if hasattr(template, 'status'):
        if template.status == 'active':
            print("   ✅ Template is ACTIVE - ready to use!")
        elif template.status == 'building':
            print("   ⏳ Template is still BUILDING - wait before using")
        elif template.status == 'publishing':
            print("   ⏳ Template is PUBLISHING - almost ready")
        elif template.status == 'failed':
            print("   ❌ Template build FAILED - cannot use")
    elif template.is_active:
        print("   ✅ Template is active (via is_active flag)")
    
    print()
    
    return template


def create_sandbox_from_template(template_name: str):
    """Create sandbox from template"""
    print("="*60)
    print(f"3. CREATE SANDBOX FROM TEMPLATE: {template_name}")
    print("="*60)
    
    # IMPORTANT: Always check template status before creating sandbox
    template = Sandbox.get_template(template_name, api_key=API_KEY)
    
    if hasattr(template, 'status') and template.status != 'active':
        print(f"❌ Template status is '{template.status}' - must be 'active'")
        print("   Wait for template to finish building before creating sandbox")
        return None
    
    sandbox = Sandbox.create(
        template=template_name,
        api_key=API_KEY
    )
    
    print(f"✅ Sandbox created from template '{template_name}'")
    print(f"   Sandbox ID: {sandbox.sandbox_id}")
    
    info = sandbox.get_info()
    print(f"   Status: {info.status}")
    print(f"   Agent URL: {info.public_host}")
    print(f"   Template ID: {info.template_id}")
    print()
    
    # Test code execution
    print("Testing code execution...")
    result = sandbox.run_code("print('Hello from', __import__('platform').platform())")
    if result.success:
        print(f"✅ Code execution test passed")
        print(f"   Output: {result.stdout.strip()}")
    else:
        print(f"❌ Code execution test failed")
        print(f"   Error: {result.stderr}")
    
    print()
    
    return sandbox


def template_status_guide():
    """Guide on template statuses"""
    print("="*60)
    print("4. TEMPLATE STATUS GUIDE")
    print("="*60)
    
    print("""
Template Lifecycle Statuses:

1. 🔨 building
   - Template is being built
   - Cannot create sandboxes yet
   - Wait for completion

2. 📦 publishing  
   - Build complete, preparing template
   - Almost ready
   - Wait a few more seconds

3. ✅ active
   - Template is ready to use
   - Can create sandboxes
   - This is what you want!

4. ❌ failed
   - Build failed
   - Cannot use this template
   - Check build logs for errors

Best Practice:
--------------
Always check template.status == 'active' before creating a sandbox:

    template = Sandbox.get_template("my-template")
    
    if template.status == 'active':
        sandbox = Sandbox.create(template="my-template")
    else:
        print(f"Wait... status is {template.status}")
""")
    print()


def main():
    """Run all examples"""
    print("\n" + "="*60)
    print("PYTHON SDK - TEMPLATE MANAGEMENT")
    print("="*60)
    print()
    
    # Example 1: List all templates
    templates = list_all_templates()
    
    # Example 2: Get template details
    template = get_template_details()
    
    # Example 3: Create sandbox from template
    sandbox = None
    try:
        sandbox = create_sandbox_from_template("code-interpreter")
    finally:
        if sandbox:
            print("Cleaning up...")
            sandbox.kill()
            print("✅ Sandbox deleted")
    
    # Example 4: Status guide
    template_status_guide()
    
    print("="*60)
    print("✅ ALL EXAMPLES COMPLETED!")
    print("="*60)
    print()


if __name__ == "__main__":
    main()

File: cookbook/javascript/11_template_management.ts
/**
 * HOPX.AI JavaScript/TypeScript SDK - Template Management
 * 
 * This example demonstrates:
 * 1. Listing templates
 * 2. Getting template details
 * 3. Creating sandbox from template
 */

import { Sandbox } from '@hopx-ai/sdk';

const API_KEY = process.env.HOPX_API_KEY || 'your-api-key-here';

if (API_KEY === 'your-api-key-here') {
  throw new Error('HOPX_API_KEY environment variable not set');
}

async function listAllTemplates() {
  console.log('='.repeat(60));
  console.log('LIST ALL TEMPLATES');
  console.log('='.repeat(60));
  
  const templates = await Sandbox.listTemplates({ apiKey: API_KEY });
  
  console.log(`Found ${templates.length} template(s):\n`);
  
  for (const template of templates) {
    console.log(`📦 ${template.name}`);
    console.log(`   ID: ${template.id}`);
    console.log(`   Display Name: ${template.displayName}`);
    if (template.category) {
      console.log(`   Category: ${template.category}`);
    }
    if (template.language) {
      console.log(`   Language: ${template.language}`);
    }
    if (template.description) {
      console.log(`   Description: ${template.description}`);
    }
    if (template.status) {
      console.log(`   Status: ${template.status}`);
    }
    console.log();
  }
  
  return templates;
}

async function getTemplateDetails() {
  console.log('='.repeat(60));
  console.log('GET TEMPLATE DETAILS');
  console.log('='.repeat(60));
  
  const templateName = 'code-interpreter';
  console.log(`Getting details for '${templateName}'...\n`);
  
  const template = await Sandbox.getTemplate(templateName, { apiKey: API_KEY });
  
  console.log(`✅ Template: ${template.name}`);
  console.log(`   Display Name: ${template.displayName}`);
  if (template.category) {
    console.log(`   Category: ${template.category}`);
  }
  if (template.language) {
    console.log(`   Language: ${template.language}`);
  }
  if (template.description) {
    console.log(`   Description: ${template.description}`);
  }
  if (template.status) {
    console.log(`   Status: ${template.status}`);
  }
  console.log();
  
  return template;
}

async function createSandboxFromTemplate(templateName: string) {
  console.log('='.repeat(60));
  console.log(`CREATE SANDBOX FROM TEMPLATE: ${templateName}`);
  console.log('='.repeat(60));
  
  const sandbox = await Sandbox.create({
    template: templateName,
    apiKey: API_KEY
  });
  
  console.log(`✅ Sandbox created from template '${templateName}'`);
  console.log(`   Sandbox ID: ${sandbox.sandboxId}`);
  
  const info = await sandbox.getInfo();
  console.log(`   Status: ${info.status}`);
  console.log(`   Agent URL: ${info.publicHost}`);
  if (info.templateId) {
    console.log(`   Template ID: ${info.templateId}`);
  }
  console.log();
  
  console.log('Testing code execution...');
  const result = await sandbox.runCode("console.log('Hello from', process.platform);");
  if (result.success) {
    console.log(`✅ Code execution test passed`);
    console.log(`   Output: ${result.stdout.trim()}`);
  } else {
    console.log(`❌ Code execution test failed`);
    console.log(`   Error: ${result.stderr}`);
  }
  
  console.log();
  
  return sandbox;
}

async function templateStatusGuide() {
  console.log('='.repeat(60));
  console.log('TEMPLATE STATUS GUIDE');
  console.log('='.repeat(60));
  console.log();
  console.log('Template lifecycle statuses:');
  console.log();
  console.log('  building   - Template is being built (Dockerfile processing)');
  console.log('  publishing - Build complete, being published to registry');
  console.log('  active     - Ready to use! ✅');
  console.log('  failed     - Build or publishing failed ❌');
  console.log();
  console.log('⚠️  Only use templates with status=\'active\'');
  console.log();
  console.log('Template.build() waits automatically for \'active\' status,');
  console.log('so you can create sandboxes immediately after building.');
  console.log();
}

async function main() {
  console.log('\n' + '='.repeat(60));
  console.log('JAVASCRIPT/TYPESCRIPT SDK - TEMPLATE MANAGEMENT');
  console.log('='.repeat(60));
  console.log();
  
  await templateStatusGuide();
  
  const templates = await listAllTemplates();
  const template = await getTemplateDetails();
  
  let sandbox: Awaited<ReturnType<typeof Sandbox.create>> | null = null;
  try {
    sandbox = await createSandboxFromTemplate('code-interpreter');
  } finally {
    if (sandbox) {
      console.log('Cleaning up...');
      await sandbox.kill();
      console.log('✅ Sandbox deleted');
    }
  }
  
  console.log('='.repeat(60));
  console.log('✅ ALL TESTS COMPLETED!');
  console.log('='.repeat(60));
  console.log();
}

main().catch(console.error);

